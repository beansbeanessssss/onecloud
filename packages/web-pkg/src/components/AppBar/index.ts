export { default as AppBar } from './AppBar.vue'
export { default as SidebarToggle } from './SidebarToggle.vue'
