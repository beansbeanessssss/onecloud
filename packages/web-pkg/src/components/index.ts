export * from './ContextActions'
export * from './sideBar'
export * from './Spaces'

export { default as AppLoadingSpinner } from './AppLoadingSpinner.vue'
export { default as AppTopBar } from './AppTopBar.vue'
export { default as BatchActions } from './BatchActions.vue'
export { default as ItemFilter } from './ItemFilter.vue'
export { default as LoadingIndicator } from './LoadingIndicator.vue'
export { default as NoContentMessage } from './NoContentMessage.vue'
export { default as QuotaSelect } from './QuotaSelect.vue'
export { default as SpaceQuota } from './SpaceQuota.vue'
