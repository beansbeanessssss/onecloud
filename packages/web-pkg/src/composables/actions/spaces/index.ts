export * from './useSpaceActionsDelete'
export * from './useSpaceActionsDisable'
export * from './useSpaceActionsDuplicate'
export * from './useSpaceActionsEditDescription'
export * from './useSpaceActionsEditQuota'
export * from './useSpaceActionsEditReadmeContent'
export * from './useSpaceActionsRename'
export * from './useSpaceActionsRestore'
export * from './useSpaceActionsShowMembers'
