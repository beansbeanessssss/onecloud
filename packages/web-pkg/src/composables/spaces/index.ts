export * from './useSpaceHelpers'
export * from './useGetMatchingSpace'
export * from './useCreateSpace'
