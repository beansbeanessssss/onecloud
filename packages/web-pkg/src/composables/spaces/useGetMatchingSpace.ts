import { useCapabilitySpacesEnabled } from '../capability'
import { useStore } from '../store'
import { useConfigurationManager } from '../configuration'
import { useRouteParam } from '../router'
import { Resource, SpaceResource } from '@ownclouders/web-client'
import {
  MountPointSpaceResource,
  ProjectSpaceResource,
  User,
  buildShareSpaceResource,
  extractStorageId,
  isMountPointSpaceResource,
  isPersonalSpaceResource,
  isProjectSpaceResource
} from '@ownclouders/web-client/src/helpers'
import { computed, Ref, unref } from 'vue'
import { basename } from 'path'

type GetMatchingSpaceOptions = {
  space?: Ref<SpaceResource>
}

export const useGetMatchingSpace = (options?: GetMatchingSpaceOptions) => {
  const store = useStore()
  const configurationManager = useConfigurationManager()
  const spaces = computed<ProjectSpaceResource[]>(() => store.getters['runtime/spaces/spaces'])
  const user = computed<User>(() => store.getters.user)
  const driveAliasAndItem = useRouteParam('driveAliasAndItem')
  const hasSpaces = useCapabilitySpacesEnabled(store)

  const getInternalSpace = (storageId: string): SpaceResource => {
    return (
      unref(options?.space) ||
      unref(spaces).find((space) => space.id === storageId) ||
      (!unref(hasSpaces) && unref(spaces).find((s) => isPersonalSpaceResource(s)))
    )
  }

  const getMatchingSpace = (resource: Resource): SpaceResource => {
    let storageId = resource.storageId

    if (unref(driveAliasAndItem)?.startsWith('public/')) {
      storageId = unref(driveAliasAndItem).split('/')[1]
    }

    return (
      getInternalSpace(storageId) ||
      buildShareSpaceResource({
        shareId: resource.shareId,
        shareName: resource.shareRoot ? basename(resource.shareRoot) : resource.name,
        serverUrl: configurationManager.serverUrl
      })
    )
  }

  const getMatchingMountPoints = (space: SpaceResource): MountPointSpaceResource[] =>
    unref(spaces).filter(
      (s) => isMountPointSpaceResource(s) && extractStorageId(s.root.remoteItem.rootId) === space.id
    )

  const isResourceAccessible = ({ space, path }: { space: SpaceResource; path: string }) => {
    if (!configurationManager.options.routing.fullShareOwnerPaths) {
      return true
    }

    const projectSpace = unref(spaces).find((s) => isProjectSpaceResource(s) && s.id === space.id)
    const fullyAccessibleSpace = space.isOwner(unref(user)) || projectSpace?.isMember(unref(user))

    return (
      fullyAccessibleSpace ||
      getMatchingMountPoints(space).some((m) => path.startsWith(m.root.remoteItem.path))
    )
  }

  return {
    getInternalSpace,
    getMatchingSpace,
    isResourceAccessible
  }
}
