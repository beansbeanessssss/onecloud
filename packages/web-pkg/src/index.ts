export * from './components'
export * from './composables'
export * from './configuration'
export * from './errors'
export * from './helpers'
export * from './http'
export * from './observer'
export * from './services'
export * from './utils'
