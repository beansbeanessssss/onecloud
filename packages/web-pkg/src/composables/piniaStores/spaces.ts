import { defineStore } from 'pinia'
import { computed, ref, unref } from 'vue'
import { Graph, SpaceResource } from '@ownclouders/web-client'
import {
  GraphShareRoleIdMap,
  buildSpace,
  extractStorageId,
  isPersonalSpaceResource,
  isProjectSpaceResource
} from '@ownclouders/web-client/src/helpers'
import type { CollaboratorShare } from '@ownclouders/web-client/src/helpers'
import { useUserStore } from './user'
import { ConfigStore, useConfigStore } from './config'
import { buildCollaboratorShare } from '@ownclouders/web-client/src/helpers/share/functionsNG'
import { useSharesStore } from './shares'
import { Permission } from '@ownclouders/web-client/src/generated'

export const sortSpaceMembers = (shares: CollaboratorShare[]) => {
  const sortedManagers = shares
    .filter((share) => share.role.id === GraphShareRoleIdMap.SpaceManager)
    .sort((a, b) => a.sharedWith.displayName.localeCompare(b.sharedWith.displayName))

  const sortedRest = shares
    .filter((share) => share.role.id !== GraphShareRoleIdMap.SpaceManager)
    .sort((a, b) => a.sharedWith.displayName.localeCompare(b.sharedWith.displayName))

  return [...sortedManagers, ...sortedRest]
}

export const getSpacesByType = async ({
  graphClient,
  driveType,
  configStore
}: {
  graphClient: Graph
  driveType: string
  configStore: ConfigStore
}) => {
  const graphResponse = await graphClient.drives.listMyDrives(
    'name asc',
    `driveType eq ${driveType}`
  )
  if (!graphResponse.data) {
    return []
  }

  const mountpoints = graphResponse.data.value.map((space) =>
    buildSpace({ ...space, serverUrl: configStore.serverUrl })
  )
  if (driveType !== 'mountpoint' || !configStore.options.routing?.fullShareOwnerPaths) {
    return mountpoints
  }

  const rootSpaceDriveAliasMapping: Record<string, string> = {}
  graphResponse.data.value.forEach((space) => {
    const { rootId, driveAlias } = space.root.remoteItem
    rootSpaceDriveAliasMapping[rootId] = driveAlias
  })

  const rootSpaces = Object.entries(rootSpaceDriveAliasMapping).map(([id, driveAlias]) =>
    // FIXME: create proper buildRootSpace (or whatever function)
    buildSpace({
      id: extractStorageId(id),
      name: driveAlias, // FIXME: set a proper name
      driveType: driveAlias.split('/')[0], // FIXME: can we retrieve this from api?
      driveAlias,
      path: '/',
      serverUrl: configStore.serverUrl
    })
  )

  return [...mountpoints, ...rootSpaces]
}

export const useSpacesStore = defineStore('spaces', () => {
  const userStore = useUserStore()
  const configStore = useConfigStore()
  const sharesStore = useSharesStore()

  const spaces = ref<SpaceResource[]>([])
  const spaceMembers = ref<CollaboratorShare[]>([])
  const currentSpace = ref<SpaceResource>()
  const spacesInitialized = ref(false)
  const mountPointsInitialized = ref(false)
  const spacesLoading = ref(false)

  const personalSpace = computed(() => {
    return unref(spaces).find((s) => isPersonalSpaceResource(s) && s.isOwner(userStore.user))
  })

  const setSpacesInitialized = (value: boolean) => {
    spacesInitialized.value = value
  }

  const setMountPointsInitialized = (value: boolean) => {
    mountPointsInitialized.value = value
  }

  const setSpacesLoading = (value: boolean) => {
    spacesLoading.value = value
  }

  const setCurrentSpace = (space: SpaceResource) => {
    currentSpace.value = space
  }

  const setSpaceMembers = (members: CollaboratorShare[]) => {
    spaceMembers.value = members
  }

  const addSpaces = (s: SpaceResource[]) => {
    unref(spaces).push(...s)
  }

  const removeSpace = (space: SpaceResource) => {
    spaces.value = unref(spaces).filter(({ id }) => id !== space.id)
  }

  const upsertSpace = (space: SpaceResource) => {
    const existingSpace = unref(spaces).find(({ id }) => id === space.id)
    if (existingSpace) {
      Object.assign(existingSpace, space)
      return
    }
    addSpaces([space])
  }

  const updateSpaceField = <T extends SpaceResource, K extends keyof SpaceResource>({
    id,
    field,
    value
  }: {
    id: T['id']
    field: K
    value: T[K]
  }) => {
    const space = unref(spaces).find((space) => id === space.id)
    if (space) {
      space[field] = value
    }
  }

  const loadSpaces = async ({ graphClient }: { graphClient: Graph }) => {
    spacesLoading.value = true
    try {
      /**
       * FIXME: this is bad for two reasons:
       * 1. fetching by specific drive type is bad because if more drive types are being added it needs additional code.
       *    as soon as the backend allows to filter by `driveType neq virtual` we want to use that here.
       * 2. fetching the mountpoint drives only on first access is kind of error prone, because mount points are
       *    trying to be accessed in multiple code locations. all of them need to check now if mountpoints need to be
       *    fetched first. but at the moment fetching mountpoints is kind of expensive, so we need to accept that for now.
       */
      const [personalSpaces, projectSpaces] = await Promise.all([
        getSpacesByType({ graphClient, driveType: 'personal', configStore }),
        getSpacesByType({ graphClient, driveType: 'project', configStore })
      ])

      addSpaces([...personalSpaces, ...projectSpaces])
      spacesInitialized.value = true
    } finally {
      spacesLoading.value = false
    }
  }

  const loadMountPoints = async ({ graphClient }: { graphClient: Graph }) => {
    // fetching mount points is particularly expensive, so we do that only on first access.
    if (unref(mountPointsInitialized)) {
      return
    }
    try {
      const mountPointSpaces = await getSpacesByType({
        graphClient,
        driveType: 'mountpoint',
        configStore
      })
      addSpaces(mountPointSpaces)
    } finally {
      mountPointsInitialized.value = true
    }
  }

  const reloadProjectSpaces = async ({ graphClient }: { graphClient: Graph }) => {
    const projectSpaces = await getSpacesByType({ graphClient, driveType: 'project', configStore })
    spaces.value = unref(spaces).filter((s) => !isProjectSpaceResource(s))
    addSpaces(projectSpaces)
  }

  const loadSpaceMembers = async ({
    graphClient,
    space
  }: {
    graphClient: Graph
    space: SpaceResource
  }) => {
    spaceMembers.value = []
    const spaceShares: CollaboratorShare[] = []

    const { data } = await graphClient.permissions.listPermissions(
      space.id,
      `${space.id}!${space.id.split('$')[1]}` // FIXME: server bug
    )

    const permissions = ((data as any).value || []) as Permission[]
    permissions.forEach((graphPermission) => {
      if (!graphPermission.link) {
        spaceShares.push(
          buildCollaboratorShare({
            graphPermission,
            graphRoles: sharesStore.graphRoles,
            resourceId: space.id,
            user: userStore.user
          })
        )
      }
    })

    spaceMembers.value = sortSpaceMembers(spaceShares)
  }

  const upsertSpaceMember = ({ member }: { member: CollaboratorShare }) => {
    const existingMember = unref(spaceMembers).find(({ id }) => id === member.id)
    if (existingMember) {
      Object.assign(existingMember, member)
      return
    }

    unref(spaceMembers).push(member)
  }

  const removeSpaceMember = ({ member }: { member: CollaboratorShare }) => {
    const existingMember = unref(spaceMembers).find(({ id }) => id === member.id)
    spaceMembers.value = unref(spaceMembers).filter(({ id }) => existingMember.id !== id)
  }

  return {
    spaces,
    spacesInitialized,
    mountPointsInitialized,
    spacesLoading,
    spaceMembers,
    currentSpace,
    personalSpace,

    setSpacesInitialized,
    setMountPointsInitialized,
    setSpacesLoading,
    setCurrentSpace,
    setSpaceMembers,

    addSpaces,
    removeSpace,
    upsertSpace,
    updateSpaceField,
    loadSpaces,
    loadMountPoints,
    reloadProjectSpaces,

    loadSpaceMembers,
    upsertSpaceMember,
    removeSpaceMember
  }
})

export type SpacesStore = ReturnType<typeof useSpacesStore>
