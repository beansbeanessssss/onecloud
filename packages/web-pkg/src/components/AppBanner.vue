<template>
  <portal to="app.app-banner">
    <div v-if="isAppBannerAvailable" class="app-banner hide-desktop" :hidden="isVisible === false">
      <oc-button
        variation="brand"
        appearance="raw"
        class="app-banner-exit"
        aria-label="Close"
        @click="close"
      >
        <oc-icon name="close" size="small" />
      </oc-button>
      <div
        class="app-banner-icon"
        :style="{ 'background-image': `url('${appBannerSettings.icon}')` }"
      ></div>
      <div class="info-container">
        <div>
          <div class="app-title">{{ appBannerSettings.title }}</div>
          <div class="app-publisher">{{ appBannerSettings.publisher }}</div>
          <div v-if="appBannerSettings.additionalInformation !== ''" class="app-additional-info">
            {{ $gettext(appBannerSettings.additionalInformation) }}
          </div>
        </div>
      </div>
      <a
        :href="appUrl"
        target="_blank"
        class="app-banner-cta"
        rel="noopener"
        aria-label="{{ $gettext(appBannerSettings.ctaText) }}"
        >{{ $gettext(appBannerSettings.ctaText) }}</a
      >
    </div>
  </portal>
</template>

<script lang="ts">
import { computed, defineComponent, ref, unref } from 'vue'
import { useRouter, useStore } from '../composables'
import { buildUrl } from '../helpers/router'
import { useSessionStorage } from '@vueuse/core'

export default defineComponent({
  components: {},
  props: {
    fileId: {
      type: String,
      required: true
    }
  },
  setup(props) {
    const appBannerWasClosed = useSessionStorage('app_banner_closed', null)
    const isVisible = ref<boolean>(unref(appBannerWasClosed) === null)
    const store = useStore()
    const router = useRouter()
    const appBannerSettings = unref(store.getters.configuration.currentTheme.appBanner)
    const isAppBannerAvailable = computed(
      () => appBannerSettings && Object.keys(appBannerSettings).length != 0
    )
    const appUrl = computed(() => {
      return buildUrl(router, `/f/${props.fileId}`)
        .toString()
        .replace('https', appBannerSettings.appScheme)
    })

    const close = () => {
      isVisible.value = false
      useSessionStorage('app_banner_closed', 1)
    }

    return {
      appUrl,
      close,
      isVisible,
      appBannerSettings,
      isAppBannerAvailable
    }
  }
})
</script>

<style scoped lang="scss">
.hide-desktop {
  @media (min-width: 768px) {
    display: none;
  }
}

.app-banner {
  overflow-x: hidden;
  width: 100%;
  height: 84px;
  background: #f3f3f3;
  font-family: Helvetica, sans, sans-serif;
  z-index: 5;
}

.info-container {
  position: absolute;
  top: 10px;
  left: 104px;
  display: flex;
  overflow-y: hidden;
  width: 60%;
  height: 64px;
  align-items: center;
  color: #000;
}

.app-banner-icon {
  position: absolute;
  top: 10px;
  left: 30px;
  width: 64px;
  height: 64px;
  border-radius: 15px;
  background-size: 64px 64px;
}

.app-banner-cta {
  position: absolute;
  top: 32px;
  right: 10px;
  z-index: 1;
  display: block;
  padding: 0 10px;
  min-width: 10%;
  border-radius: 5px;
  background: #f3f3f3;
  color: #1474fc;
  font-size: 18px;
  text-align: center;
  text-decoration: none;
}

.app-title {
  font-size: 14px;
}

.app-publisher,
.app-additional-info {
  font-size: 12px;
}

.app-banner-exit {
  position: absolute;
  top: 34px;
  left: 9px;
  margin: 0;
  width: 12px;
  height: 12px;
  border: 0;
  text-align: center;
  display: inline;
}
</style>
