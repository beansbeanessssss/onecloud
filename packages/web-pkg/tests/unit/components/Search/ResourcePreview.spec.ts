import { mock } from 'jest-mock-extended'
import { ResourcePreview } from '../../../../src/components'
import { SpaceResource } from '@ownclouders/web-client/src'
import { useGetMatchingSpace } from '../../../../src/composables/spaces/useGetMatchingSpace'
import {
  createStore,
  defaultComponentMocks,
  defaultPlugins,
  shallowMount,
  defaultStoreMockOptions,
  useGetMatchingSpaceMock
} from 'web-test-helpers'

jest.mock('../../../../src/composables/spaces/useGetMatchingSpace', () => ({
  useGetMatchingSpace: jest.fn()
}))

describe('Preview component', () => {
  const driveAliasAndItem = '1'
  jest.mocked(useGetMatchingSpace).mockImplementation(() => useGetMatchingSpaceMock())
  it('should render preview component', () => {
    const { wrapper } = getWrapper({
      space: mock<SpaceResource>({
        id: '1',
        driveType: 'project',
        name: 'New space',
        getDriveAliasAndItem: () => driveAliasAndItem
      })
    })
    expect(wrapper.html()).toMatchSnapshot()
  })
})

function getWrapper({
  route = {
    query: {},
    params: {}
  },
  hasShareJail = true,
  space = null,
  searchResult = {
    id: '1',
    data: {
      storageId: '1',
      name: 'lorem.txt',
      path: '/',
      shareRoot: ''
    }
  },
  user = { id: 'test' }
}: {
  route?: any
  hasShareJail?: boolean
  space?: SpaceResource
  searchResult?: any
  user?: any
} = {}) {
  jest.mocked(useGetMatchingSpace).mockImplementation(() =>
    useGetMatchingSpaceMock({
      isResourceAccessible() {
        return true
      },
      getMatchingSpace() {
        return space
      }
    })
  )

  const storeOptions = {
    ...defaultStoreMockOptions,
    getters: {
      ...defaultStoreMockOptions.getters,
      configuration: () => ({
        options: {
          disablePreviews: true
        }
      }),
      capabilities: () => ({
        spaces: {
          share_jail: hasShareJail,
          projects: { enabled: true }
        }
      }),
      user: () => user
    }
  }
  const store = createStore(storeOptions)
  const mocks = defaultComponentMocks({ currentRoute: route })
  return {
    wrapper: shallowMount(ResourcePreview, {
      props: {
        searchResult
      },
      global: {
        provide: mocks,
        renderStubDefaultSlot: true,
        mocks,
        plugins: [...defaultPlugins(), store]
      }
    })
  }
}
