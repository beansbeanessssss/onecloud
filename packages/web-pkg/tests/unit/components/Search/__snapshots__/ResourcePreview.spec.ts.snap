// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Preview component should render preview component 1`] = `<oc-resource-stub folderlink="[object Object]" isextensiondisplayed="true" isicondisplayed="true" ispathdisplayed="true" isresourceclickable="true" isthumbnaildisplayed="false" parentfolderlink="[object Object]" parentfolderlinkiconadditionalattributes="[object Object]" parentfoldername="New space" pathprefix="Spaces/New space" resource="[object Object]"></oc-resource-stub>`;
